streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java

```diff
@@ -305,14 +306,16 @@ private void recordSendError(final String topic, final Exception exception, fina
                 "indicating the task may be migrated out";
             sendException.set(new TaskMigratedException(errorMessage, exception));
         } else {
-            if (exception instanceof RetriableException) {
+            final ProductionExceptionHandlerResponse response = productionExceptionHandler.handle(serializedRecord, exception);
```

Comment:
I don't think we need to call the handler for all cases here, but only if the exception is non-retriable (old, existing condition), or if the exception is a timeout with root cause "unknown topic or partition"?

So something like:
```
            if (!isRetriable(exception)) { // we only change this line but the rest stays the same?
errorMessage += "\nThe broker is either slow or in bad state (like not having enough replicas) in responding the request, " +
                    "or the connection to broker was interrupted sending the request or receiving the response. " +
                    "\nConsider overwriting `max.block.ms` and /or " +
                    "`delivery.timeout.ms` to a larger value to wait longer for such scenarios and avoid timeout errors";
                sendException.set(new TaskCorruptedException(Collections.singleton(taskId)));
            } else {
                if (productionExceptionHandler.handle(serializedRecord, exception) == ProductionExceptionHandlerResponse.FAIL) {
                    errorMessage += "\nException handler choose to FAIL the processing, no more records would be sent.";
                    sendException.set(new StreamsException(errorMessage, exception));
                } else {
                    errorMessage += "\nException handler choose to CONTINUE processing in spite of this error but written offsets would not be recorded.";
                    droppedRecordsSensor.record();
                }
```

Reply:
If the exception is `Timeout` with root cause `UnknownTopicOrPartition` and the handler instructs to `CONTINUE`, we must follow the old logic, which is having  `TaskCorruptedException`. Therefore, I think we have to call handler for nearly all cases.